Sming/SmingCore/Data/ObjectMap.h

```diff
@@ -0,0 +1,249 @@
+	 * @brief Get the number of entries in this map
+	 * @retval int Entry count
+	 */
+	unsigned count() const
+	{
+		return entries.count();
+	}
+
+	/*
+	 * @brief Get a key at a specified index, non-modifiable
+	 * @param idx the index to get the key at
+	 * @return The key at index idx
+	 */
+	const K& keyAt(unsigned idx) const
+	{
+		return entries[idx].key;
+	}
+
+	/*
+	 * @brief Get a key at a specified index
+	 * @param idx the index to get the key at
+	 * @return Reference to the key at index idx
+	 */
+	K& keyAt(unsigned idx)
+	{
+		return entries[idx].key;
+	}
+
+	/*
+	 * @brief Get a value at a specified index, non-modifiable
+	 * @param idx the index to get the value at
+	 * @retval The value at index idx
+	 * @note The caller must not use `delete` on the returned value
+	 */
+	const V* valueAt(unsigned idx) const
+	{
+		return entries[idx].value;
+	}
+
+	/*
+	 * @brief Get a value at a specified index
+	 * @param idx the index to get the value at
+	 * @retval The value at index idx
+	 * @note Because a reference is returned any existing value must be `delete`d first
+	 * @see `operator[]`
+	 * @see `set()`
+	 */
+	V*& valueAt(unsigned idx)
+	{
+		return entries[idx].value;
+	}
+
+	/**
+	 * @brief Get value for given key, if it exists
+	 * @param key
+	 * @retval const V* Will be null if not found in the map
+	 */
+	const V* operator[](const K& key) const
+	{
+		return find(key);
+	}
+
+	/** @brief Access map entry by reference
+	 *  @param key
+	 *  @retval V*& Reference to mapped value corresponding to given key
+	 *  @note If the given key does not exist in the map then it will be created
+	 *  and a null value entry returned.
+	 *  Be careful to check for existing value before assigning to avoid memory leaks.
+	 *  In most cases `set()` should be used to avoid this complication.
+	 *
+	 *  Example:
+	 *
+	 *  ```
+	 *	void test()
+	 *	{
+	 *		ObjectMap<String, MyType> map;
+	 *		MyType* object1 = new MyType();
+	 *		MyType* object2 = new MyType();
+	 *		map["key1"] = object1;
+	 *		auto& value = map["key1"]; // value now refers to object1
+	 *		delete value;
+	 *		value = object2;
+	 *		// As soon as `map` goes out of scope, object2 is released
+	 *	}
+	 * 	```
+	 *
+	 * 	@see `set()`
+	 * 	@see `valueAt()`
+	 *
+	 */
+	V*& operator[](const K& key)
+	{
+		int i = indexOf(key);
+		if(i >= 0) {
+			return entries[i].value;
+		}
+
+		auto entry = new Entry(key, nullptr);
+		entries.addElement(entry);
+		return entry->value;
+	}
+
+	/**
+	 * @brief Set a key value, ensuring any existing value is released
+	 *  @param key
+	 *  @param value
+	 *  @note Example:
+	 *  Example:
+	 *
+	 *		void test()
+	 *		{
+	 *			ObjectMap<String, MyType> map;
+	 *			MyType* object1 = new MyType();
+	 *			MyType* object2 = new MyType();
+	 *			map.set("key1", object1);
+	 *			map.set("key1", object2); // object1 is freed automatically
+	 *			// As soon as `map` goes out of scope, object2 is released
+	 *		}
+	 *
+	 */
+	void set(const K& key, V* value)

```

Comment:
That `set` will cause confusion. We have two ways to add things to the map that behave diffently:

```
ObjectMap<String, MyType> map;
map["key1"] = object1;
// and 
map.set("key1, object1);
```


The best would have been to make `map["key1"] = object1;` behave as `set`.  But probably it would be better to remove completely `V& operator[](const K& key)` since I cannot think of a proper way to implement the above.

Reply:
> The best would have been to make `map["key1"] = object1;` behave as `set`

Done :-)